data/sqlutil/macros.go

```diff
@@ -123,119 +125,96 @@ var DefaultMacros = Macros{
 	"column":     macroColumn,
 }
 
-func trimAll(s []string) []string {
-	r := make([]string, len(s))
-	for i, v := range s {
-		r[i] = strings.TrimSpace(v)
-	}
-
-	return r
+type macroMatch struct {
+	full string
+	args []string
 }
 
 // getMacroMatches extracts macro strings with their respective arguments from the sql input given
 // It manually parses the string to find the closing parenthesis of the macro (because regex has no memory)
-func getMacroMatches(input string, name string) ([][]string, error) {
-	macroName := fmt.Sprintf("\\$__%s\\b", name)
-	matchedMacros := [][]string{}
-
-	rgx, err := regexp.Compile(macroName)
+func getMacroMatches(input string, name string) ([]macroMatch, error) {
+	rgx, err := regexp.Compile(fmt.Sprintf(`\$__%s\b`, name))
 	if err != nil {
 		return nil, err
 	}
 
-	// get all matching macro instances
-	matched := rgx.FindAllStringIndex(input, -1)
-	if matched == nil {
-		return nil, nil
+	var matches []macroMatch
+	for _, window := range rgx.FindAllStringIndex(input, -1) {
+		start, end := window[0], window[1]
+		args, length := parseArgs(input[end:])
+		if length < 0 {
+			return nil, fmt.Errorf("failed to parse macro arguments (missing close bracket?)")
+		}
+		matches = append(matches, macroMatch{full: input[start : end+length], args: args})
+	}
+	return matches, nil
+}
+
+// parseArgs looks for a bracketed argument list at the beginning of
+// argString. If one is present, returns a list of whitespace-trimmed
+// arguments and the length of the string comprising the argument list.
+func parseArgs(argString string) ([]string, int) {
+	if argString == "" || argString[0] != '(' {
+		return nil, 0
 	}
 
-	for matchedIndex := 0; matchedIndex < len(matched); matchedIndex++ {
-		var macroEnd = 0
-		var argStart = 0
-		// quick exit from the loop, when we encounter a closing bracket before an opening one (ie "($__macro)", where we can skip the closing one from the result)
-		var forceBreak = false
-		macroStart := matched[matchedIndex][0]
-		inputCopy := input[macroStart:]
-		cache := make([]rune, 0)
-
-		// find the opening and closing arguments brackets
```

Comment:
In `sqlds` the logic here around parsing macro arguments is partially duplicated: once in `getMacroMatches` to find the range of the query string that represents arguments and once in `parseArgs` to extract those arguments. This logic was implemented slightly differently in both places but `parseArgs` was effectively doing all the work.